File: 이코테 복습(python, c++, java)/이진탐색/c++/정렬된 배열에서 특정 수의 개수 구하기.cpp
#include<iostream>
#include<algorithm>
#include<vector>
using namespace std;

int countByRange(vector<int>& v, int leftValue, int rightValue) {
	vector<int>::iterator rightIndex = upper_bound(v.begin(), v.end(), rightValue);
	vector<int>::iterator leftIndex = upper_bound(v.begin(), v.end(), leftValue);
	return rightIndex - leftIndex;
}

int n, x;
vector<int> v;

int main(void) {
	cin >> n >> x;

	for (int i = 0; i < n; i++) {
		int temp;
		cin >> temp;
		v.push_back(temp);
	}

	int cnt = countByRange(v, x, x);

	if (cnt == 0) {
		cout << -1 << endl;
	}
	else {
		cout << cnt << endl;
	}
}

File: 이코테 복습(python, c++, java)/이진탐색/java/정렬된_배열에서_특정_수의_개수_구하기.java
package javastudy.이코테.이진탐색;

import java.util.Scanner;

public class 정렬된_배열에서_특정_수의_개수_구하기 {

    public static int lowerBound(int[] arr, int target, int start, int end) {
        while (start < end) {
            int mid = (start + end) / 2;
            if (arr[mid] >= target) {
                end = mid;
            } else {
                start = mid + 1;
            }
        }
        return end;
    }

    public static int upperBound(int[] arr, int target, int start, int end) {
        while (start < end) {
            int mid = (start + end) / 2;
            if (arr[mid] > target) {
                end = mid;
            } else {
                start = mid = 1;
            }
        }
        return end;
    }

    public static int countByRange(int[] arr, int leftValue, int rightValue) {
        int rightIndex = upperBound(arr, rightValue, 0, arr.length);
        int leftIndex = lowerBound(arr, leftValue, 0, arr.length);
        return rightIndex - leftIndex;
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        int n = sc.nextInt();
        int x = sc.nextInt();

        int[] arr = new int[n];
        for (int i = 0; i < n; i++) {
            arr[i] = sc.nextInt();
        }

        int cnt = countByRange(arr, x, x);

        if (cnt == 0) {
            System.out.println(-1);
        } else {
            System.out.println(cnt);
        }
    }
}
